CIP-0014/CIP-0014.md

```diff
@@ -0,0 +1,116 @@
+# CIP-0014 User-Facing Asset Id 
+
+# Preamble
+
+---
+CIP: 0014
+Title: User-Facing Asset Id
+Authors: Matthias Benkort <matthias.benkort@iohk.io>, Rodney Lorrimar <rodney.lorrimar@iohk.io>
+Comments-Summary: A common format for displaying a unique identifier for assets.
+Comments-URI: https://github.com/cardano-foundation/CIPs/wiki/Comments:CIP-0014
+Status: Draft
+Type: Standards
+Created: 2020-02-01
+License: CC-BY-4.0
+---
+
+# Abstract
+
+This specification defines a unique user-facing asset id as a bech32-encoded blake2b-256 digest of the concatenation of the policy id and the asset name.
+
+# Motivation
+
+The Mary era of Cardano introduces the support for native assets. On the blockchain, native assets are uniquely identified by both their so-called policy id and asset name. Neither the 
+policy id nor the asset name are intended to be human-readable data. 
+
+On the one hand, the policy id is a hash digest of either a monetary script or a Plutus script. On the other hand, the asset name is an arbitrary bytestring of up to 32 bytes (which does not necessarily decode to a valid UTF-8 sequence). In addition, it is possible for an asset to have an empty asset name, or, for assets to have identical asset names under different policy ids. 
+
+Because assets are manipulated in several user-facing features, and also via hardware devices, it is useful to come up with a short(er) and human-readable identifier for assets that user can recognize and refer to when talking about assets.
+
+# Specification
+
+We define a unique identifier for assets in pseudo-code as:
+
+```
+assetId := encodeBech32
+  ( datapart = hash
+    ( algorithm = 'blake2b'
+    , digest-length = 32
+    , message = policyId | assetName
+    )
+  , humanReadablePart = 'token'
+  )
+```
+
+where `|` designates the concatenation of two bytestrings and the `digest-length` is given in `bytes`.
+
+# Rationale
+
+- The asset id needs to be unique and must therefore include both the policy id and the asset name.
+- Using a hash gives us asset id of a same deterministic length which is short enough to display reasonably well on small screens.
```

Comment:
a drawback of this, probably worth acknowledging, is that the bech32 assetId, as opposed to the rest of bech32 encoded entities in Cardano as of now, cannot be easily mapped back to the elements that are getting serialized into the transaction without additional context. So tools like cardano-cli may not be able to fully adopt the bech32 id of assets (i.e. still requiring the raw policyId/assetName) if they need to work in an airgapped environment (i.e. no access to a live node database and optimized to perform the reverse search by assetId).

Reply:
That's another argument against hashing it. It may be more interoperable to simply encode the concatenation.

Reply:
yes, though also the concatenation of policyId + assetName can look confusing on hw wallets because the asset name would be apparent only after scrolling so it can lead to "pilot induced" errors when dealing with assets minted by the same policy but with a completely different purpose/value.

The assetName + policyId concatenation on the other hand can obviously encourage phishing as malicious minters can construct asset names to match legitimate assets with short names with a large part of their policy id.

So it's tricky after all and I don't see a clear winner out of these three options. Prepending a checksum, let's say CRC32 before the policyId + assetName would solve the problem outlined but it's a deviation and ultimately makes the bech32 assetId even longer

Reply:
> The assetName + policyId concatenation on the other hand can obviously encourage phishing as malicious minters can construct asset names to match legitimate assets with short names with a large part of their policy id.

I am not sure to understand what sort of phishing attempt you imagine? Are you saying that because the policyId is first, the final assetId may ressemble another one with a short asset name? Like:

`c3aa9a06767c218e51fad32996af78d67548336980c17b6c9ab166d7858421e3|bb`
`c3aa9a06767c218e51fad32996af78d67548336980c17b6c9ab166d7858421e3|66`

Which would encode to bech32 as:

`token1cw4f5pnk0sscu5066v5edtmc6e65svmfsrqhkmy6k9nd0pvyy83hewcn99vwx`
`token1cw4f5pnk0sscu5066v5edtmc6e65svmfsrqhkmy6k9nd0pvyy83hcesd82a92` 

They look indeed quite the same. That could potentially be solved by having the assetName first, and the policy id following. Hashing also obviously solves this.

Reply:
I meant that having the asset name first can encourage "phishing" in the sense that if you have the asset name let's say `b6` and the following concatenation with policyId:

```
bb|c3aa9a06767c218e51fad32996af78d67548336980c17b6c9ab166d7858421e3
```

A malicious party can create an asset with the name `bbc3aa9a06767c218e51fad32996af78d67548336980c17b6c9` which in combination with their policy id would be longer but they can "bet" on users checking just the first few characters of the given asset or some UIs performing ellipsis on longer strings. So an attacker can match the first up to 32 bytes of whatever asset they want - not necessarily an issue for HW wallets but it can mess e.g. with block explorers listing tokens and performing ellipsis on their name or complicate searchability as people tento to look things up by their first few characters

Reply:
The assetName (as long as it is human readable) should be shown on the hw-wallet displays, and not a long hex code IMO for a better user experience. There is no way to resolve this to a readable registered name, because this must also work in complete offline mode.

Reply:
>  should be shown on the hw-wallet displays,

Actually no. This opens gate for a whole class of phishing attacks. There's nothing preventing me from creating a token with `ADA` as asset name under my own policy. That would be disastrous to show to end users. The `asset_name`, despite its name **must never be shown to end users**.